Sources/NIO/EventLoopFuture.swift

```diff
@@ -217,6 +217,34 @@ public struct EventLoopPromise<Value> {
         self._resolve(value: result)
     }
 
+    /// Execute the given closure and complete the promise.
+    ///
+    /// The given closure is executed _synchronously_ or _asynchronously_ depending on whether
+    /// the current `NIOThread` is tied to the `EventLoop` that was used to create the promise.
+    ///
+    /// WARNING: You _must not_ call into any blocking code in the given closure if it is executed synchronously.
+    ///
+    /// - parameters:
+    ///     - body: The closure to execute.
+    @inlinable
+    public func completeWithClosure(_ body: @escaping () throws -> Value) {
+        func executeAndComplete() {
+            do {
+                self.succeed(try body())
+            } catch let e {
+                self.fail(e)
+            }
+        }
+
+        if self.futureResult.eventLoop.inEventLoop {
```

Comment:
I'm still unsure whether we need this. This seems to me to be conflating two different actions: one is promise completion, one is where the closure executes. I think I'd like them to be separate.

Reply:
It's a good point. I have just created another PR#1747, which removes the conflating, and also avoids the subtle behaviour change. Feel free let me know if you have any comments, thanks! 😄